File: web-dashboard/src/domain/payments/revenue.ts
// Types & helpers du **chiffre d'affaires** jour / semaine / mois — couche domaine
// (hexagonal, ADR-0008), TypeScript pur, testable sans React. **Parité stricte**
// avec le backend (`coiflink_api/domain/revenue.py`, US-6.2 #40) : pour une date de
// référence (jour civil Africa/Abidjan), le CA du salon sur **trois périodes** — le
// **jour**, la **semaine** (lundi→dimanche) et le **mois** civils qui la contiennent.
//
// Le backend reste l'**autorité des chiffres** (calcul `SUM` en base, net des
// corrections #34) : ce module ne fait que **projeter** la réponse et **formater**
// l'affichage. Les totaux sont portés en **chaîne décimale** (`NUMERIC(12,2)`) pour
// ne pas perdre de précision via un flottant JavaScript. Aucun secret, aucune PII.

import { formatXof } from "@/src/domain/payments/payment";

// Devise unique du MVP (mono-devise XOF/FCFA — PRD §9.6).
export { DEFAULT_CURRENCY } from "@/src/domain/payments/payment";

// CA d'**une** période : bornes de jour civil inclusives + total net (chaîne
// décimale) ; `total` **peut être négatif** si les corrections excèdent les
// paiements sur la période (parité backend).
export interface RevenuePeriodTotal {
  // Bornes ISO "YYYY-MM-DD" (Africa/Abidjan).
  dateFrom: string;
  dateTo: string;
  // Montant décimal en chaîne (parité `NUMERIC(12,2)`), p. ex. "35000.00".
  total: string;
}

// CA du salon sur les trois périodes pour une date de référence (miroir de
// `RevenueSummary` du backend). Objet-valeur **sans PII** : uniquement des dates,
// des montants et une devise.
export interface RevenueSummary {
  // Date de référence, ISO "YYYY-MM-DD" (Africa/Abidjan).
  referenceDate: string;
  currency: string; // «XOF» au MVP
  day: RevenuePeriodTotal;
  week: RevenuePeriodTotal;
  month: RevenuePeriodTotal;
}

// Formate le **total** d'une période en FCFA lisible (réutilise `formatXof` — le
// franc CFA s'affiche sans décimale, séparateur d'espace ; la valeur transportée
// reste la chaîne décimale d'origine).
export function formatRevenueTotal(period: RevenuePeriodTotal): string {
  return formatXof(period.total);
}

// Formate les bornes d'une période en légende compacte `JJ/MM/AAAA → JJ/MM/AAAA`.
// Une période d'un seul jour (le **jour**, `dateFrom == dateTo`) n'affiche qu'une
// date. Présentation uniquement ; jamais de recalcul de la donnée.
export function formatPeriodRange(period: RevenuePeriodTotal): string {
  const from = formatIsoDateFr(period.dateFrom);
  if (period.dateFrom === period.dateTo) return from;
  return `${from} → ${formatIsoDateFr(period.dateTo)}`;
}

// Formate une date ISO "YYYY-MM-DD" en `JJ/MM/AAAA` (fr-FR), sans dépendre du fuseau
// du navigateur (découpage textuel, pas de `new Date()`). Chaîne mal formée renvoyée
// telle quelle (défensif).
function formatIsoDateFr(iso: string): string {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso);
  if (!match) return iso;
  const [, year, month, day] = match;
  return `${day}/${month}/${year}`;
}

File: backend/coiflink_api/domain/revenue.py
"""Chiffre d'affaires **jour / semaine / mois** d'un salon (domaine pur, US-6.2, #40).

Ce module porte les objets-valeur de lecture du CA (`RevenuePeriodTotal`,
`RevenueSummary`) et — surtout — les **fonctions pures de bornes de période**
(`day_bounds` / `week_bounds` / `month_bounds`). Ces bornes **sont** la règle
métier de l'US-6.2 : « CA journalier, hebdomadaire, mensuel » se traduit par la
définition **exacte** du jour, de la semaine et du mois civils qui contiennent une
date de référence. Les isoler ici (sans I/O) les rend testables sans base et
garantit des périodes cohérentes.

Conformément à l'hexagonal (ADR-0008), le module ne connaît ni FastAPI ni
SQLAlchemy : le cas d'usage (`application/revenue.py`) convertit ces bornes de
**jour civil** en bornes UTC via `domain/time_window.py` puis interroge le journal
de caisse ; l'adapter entrant les projette en JSON.

**Sémantique des périodes** (visible par le gérant, cf. spec §Open Questions 4/9) :

- **jour** : la date de référence elle-même (`from == to`) ;
- **semaine** : la semaine civile **lundi → dimanche** qui la contient (usage
  FR/CI, standard ISO du lundi) ;
- **mois** : le mois civil (du 1er au dernier jour, via `calendar.monthrange` —
  gère février bissextile et les mois de 30/31 jours).

Ces bornes vérifient par construction `date_from ≤ date_to`, `semaine ⊇ jour` et
`mois ⊇ jour`. Elles se **chevauchent** (le jour ⊂ la semaine, la semaine croise
parfois deux mois) : le cas d'usage ne combine donc pas les sommes — il calcule
**trois** totaux indépendants.
"""

from __future__ import annotations

import calendar
import datetime
import decimal
from dataclasses import dataclass

from coiflink_api.domain.payment import DEFAULT_CURRENCY


@dataclass(frozen=True)
class RevenuePeriodTotal:
    """CA d'**une** période (jour, semaine ou mois) — projection de lecture immuable.

    Porte ses **bornes de jour civil** inclusives (`date_from`/`date_to`), le
    **total** encaissé net (`Decimal` quantifié au centime, `NUMERIC(12,2)`, jamais
    un flottant) et la **devise** (mono-devise XOF au MVP, §9.6). `total` **peut être
    négatif** si les corrections (#34) excèdent les paiements sur la période. Sans
    PII : uniquement des dates, un montant et une devise (§11.3).
    """

    date_from: datetime.date
    date_to: datetime.date
    total: decimal.Decimal
    currency: str = DEFAULT_CURRENCY


@dataclass(frozen=True)
class RevenueSummary:
    """CA d'un salon sur **trois périodes** pour une date de référence (US-6.2, #40).

    Assemble les totaux du **jour**, de la **semaine** (lundi→dimanche) et du **mois**
    civils contenant `reference_date`. Objet-valeur **immuable** et **sans PII**
    (§11.3) : seulement des dates, des montants et une devise. Construit par
    `application/revenue.py::SummarizeRevenue`.
    """

    reference_date: datetime.date
    day: RevenuePeriodTotal
    week: RevenuePeriodTotal
    month: RevenuePeriodTotal
    currency: str = DEFAULT_CURRENCY


def day_bounds(reference: datetime.date) -> tuple[datetime.date, datetime.date]:
    """Bornes du **jour** civil : la date de référence elle-même (`(d, d)`)."""

    return reference, reference


def week_bounds(reference: datetime.date) -> tuple[datetime.date, datetime.date]:
    """Bornes de la **semaine** civile **lundi → dimanche** contenant `reference`.

    `weekday()` vaut `0` le lundi et `6` le dimanche : reculer de `weekday()` jours
    donne le lundi, avancer de 6 jours depuis ce lundi donne le dimanche. Un lundi
    est donc sa propre borne basse ; un dimanche, sa propre borne haute.
    """

    monday = reference - datetime.timedelta(days=reference.weekday())
    sunday = monday + datetime.timedelta(days=6)
    return monday, sunday


def month_bounds(reference: datetime.date) -> tuple[datetime.date, datetime.date]:
    """Bornes du **mois** civil : du 1er au dernier jour du mois de `reference`.

    Le dernier jour est fourni par `calendar.monthrange` (stdlib, aucune dépendance
    nouvelle) : il gère correctement février (28/29 jours) et les mois de 30/31 jours.
    """

    first = reference.replace(day=1)
    last_day = calendar.monthrange(reference.year, reference.month)[1]
    last = reference.replace(day=last_day)
    return first, last


__all__ = [
    "RevenuePeriodTotal",
    "RevenueSummary",
    "day_bounds",
    "week_bounds",
    "month_bounds",
]
